nixos/prosody: make module defaults comply with XEP-0423

picnoir opened the pull request:
###### Motivation for this change
The overall idea behind all these changes is to provide some sane defaults to make the module "just work" by default. We are trying to stay as close as possible from the meta XEP-0423.

It is now possible to be fully complient with this meta-XEP with a minimal setup.

As an illustration, I use the following configuration on my personal prosody instance to get fully compliant to the conversations test (https://compliance.conversations.m/server/chat.alternativebit.fr/):

```nix
prosody = {
      enable = true;
      admins = [ "...@chat.alternativebit.fr" ];
      ssl.cert = "/var/lib/acme/home.alternativebit.fr/fullchain.pem";
      ssl.key = "/var/lib/acme/home.alternativebit.fr/key.pem";
      virtualHosts."chat.alternativebit.fr" = {
        enabled = true;
        domain = "chat.alternativebit.fr";
        ssl.cert = "/var/lib/acme/home.alternativebit.fr/fullchain.pem";
        ssl.key = "/var/lib/acme/home.alternativebit.fr/key.pem";
      };
      muc = [ {
        domain = "conference.alternativebit.fr";
      }];
      uploadHttp = {
        domain = "upload.alternativebit.fr";
      };
    };
  };
```

After seeing a few prosody nixos config in the wild, I'm confident this PR will make setting up prosody much easier for a lot of people.

The NixOS test has been updated to test this compliance through a VM test.

See the commit messages for more details.

----------------------
Inlined commit messages:

```
Setting up a XMPP chat server is a pretty deep rabbit whole to jump in
when you're not familiar with this whole universe. Your experience
with this environment will greatly depends on whether or not your
server implements the right set of XEPs.

To tackle this problem, the XMPP community came with the idea of
creating a meta-XEP in charge of listing the desirable XEPs to comply
with. This meta-XMP is issued every year under an new XEP number. The
2020 one being XEP-0423[1].

This prosody nixos module refactoring makes complying with XEP-0423
easier. All the necessary extensions are enabled by default. For some
extensions (MUC and HTTP_UPLOAD), we need some input from the user and
cannot provide a sensible default nixpkgs-wide. For those, we guide
the user using a couple of assertions explaining the remaining manual
steps to perform.

We took advantage of this substential refactoring to refresh the
associated nixos test.

Changelog:
- Update the prosody package to provide the necessary community
  modules in order to comply with XEP-0423. This is a tradeoff, as
  depending on their configuration, the user might end up not using them
  and wasting some disk space. That being said, adding those will
  allow the XEP-0423 users, which I expect to be the majority of
  users, to leverage a bit more the binary cache.
- Add a muc submodule populated with the prosody muc defaults.
- Add a http_upload submodule in charge of setting up a basic http
  server handling the user uploads. This submodule is in is
  spinning up an HTTP(s) server in charge of receiving and serving the
  user's attachments.
- Advertise both the MUCs and the http_upload endpoints using mod disco.
- Use the slixmpp library in place of the now defunct sleekxmpp for
  the prosody NixOS test.
- Update the nixos test to setup and test the MUC and http upload
  features.
- Add a couple of assertions triggered if the setup is not xep-0423
  compliant.

[1] https://xmpp.org/extensions/xep-0423.html
```

```
We are leveraging the systemd sandboxing features to prevent the
service accessing locations it shouldn't do. Most notably, we are here
preventing the prosody service from accessing /home and providing it
with a private /dev and /tmp.

Please consult man systemd.exec for further informations.
```


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

CC: @fpletz @globin @andir @florianjacob @flokli

picnoir committed:
nixos/prosody: make defaults comply with XEP-0423

Setting up a XMPP chat server is a pretty deep rabbit whole to jump in
when you're not familiar with this whole universe. Your experience
with this environment will greatly depends on whether or not your
server implements the right set of XEPs.

To tackle this problem, the XMPP community came with the idea of
creating a meta-XEP in charge of listing the desirable XEPs to comply
with. This meta-XMP is issued every year under an new XEP number. The
2020 one being XEP-0423[1].

This prosody nixos module refactoring makes complying with XEP-0423
easier. All the necessary extensions are enabled by default. For some
extensions (MUC and HTTP_UPLOAD), we need some input from the user and
cannot provide a sensible default nixpkgs-wide. For those, we guide
the user using a couple of assertions explaining the remaining manual
steps to perform.

We took advantage of this substential refactoring to refresh the
associated nixos test.

Changelog:
- Update the prosody package to provide the necessary community
  modules in order to comply with XEP-0423. This is a tradeoff, as
  depending on their configuration, the user might end up not using them
  and wasting some disk space. That being said, adding those will
  allow the XEP-0423 users, which I expect to be the majority of
  users, to leverage a bit more the binary cache.
- Add a muc submodule populated with the prosody muc defaults.
- Add a http_upload submodule in charge of setting up a basic http
  server handling the user uploads. This submodule is in is
  spinning up an HTTP(s) server in charge of receiving and serving the
  user's attachments.
- Advertise both the MUCs and the http_upload endpoints using mod disco.
- Use the slixmpp library in place of the now defunct sleekxmpp for
  the prosody NixOS test.
- Update the nixos test to setup and test the MUC and http upload
  features.
- Add a couple of assertions triggered if the setup is not xep-0423
  compliant.

[1] https://xmpp.org/extensions/xep-0423.html

picnoir committed:
nixos/prosody: leverage systemd sandbox features to harden service

We are leveraging the systemd sandboxing features to prevent the
service accessing locations it shouldn't do. Most notably, we are here
preventing the prosody service from accessing /home and providing it
with a private /dev and /tmp.

Please consult man systemd.exec for further informations.

NinjaTrappeur commented:
@GrahamcOfBorg test prosody

NinjaTrappeur commented:
I have a remaining question: I'd like to provide the user with an example let's encrypt setup. 

Something along the lines of

```
security.acme = {
    email = "your@email.com";
    acceptTerms = true;
    certs = {
      "your.domain.com" = {
        webroot = "/var/www/your.domain.com";
        email = "your@email.com";
        allowKeysForGroup = true;
        group = "sslcertgr";
        extraDomains."chat.domain.com" = null;
        extraDomains."upload.domain.com" = null;
        extraDomains."conference.domain.com" = null;
      };
    };
  };
```

Where should I do this? I first thought about adding an example setup in the nixos manual, but I'm not sure people will end up reading it. Do you have any idea about were to put that?

---------------

Side note: I just realized @mmilata has been bumping prosody to 0.11.5 in the meantime (took me quite some time to fix the nixos test)! Will remove this commit in a sec.

flokli commented:
Thanks a lot for doing this! I really like how this also makes the prosody module much more composable.

It's now also very easy to define mucs, which can be reused by things like https://github.com/NixOS/nixpkgs/pull/82920.

> Where should I put this?
You might just want to add a `nixos/modules/services/networking/prosody.xml` containing some documentation, as well as the examples you posted in this PR.

flokli commented on nixos/modules/services/networking/prosody.nix:
```diff
@@ -387,6 +626,27 @@ in
+        default = null;
+        example = {
+          domain = "uploads.my-xmpp-example-host.org";
+          httpsCertificate = "/etc/letsencrypt/uploads.my-xmpp-example-host.org/cert.pem";
```
shouldn't this be `/var/lib/acme/uploads.my-xmpp-example-host.org/cert.pem`?

NinjaTrappeur replied:
Ah! Good catch!

This example input was a WIP residue. When I started writing this PR, I first thought about using a TLS certificate per module (ie. one for the main VHost, one for the MUC, one for the HTTP upload server). Provided everything is then deployed on the same machine, it does not make any sense. I then decided to leverage the LE `extraDomains` feature to provide a multi domain single certificate.

picnoir committed:
nixos/prosody: add NixOS manual entry

We add a Prosody entry to the NixOS manual showing how to setup a
basic XEP-0423 compliant Prosody service. This example also showcase
how to generate the associated ACME certificates.

Note: The <programlisting> body might look poorly indented, but trust
me, it's necessary. If we try to increase their indentation level, the
HTML output will end up containing a lot of unecesseray heading spaces
breaking the formatting...

NinjaTrappeur commented:
As per suggested by @flokli, I added a Prosody section to the NixOS manual in df52c29 explaining how to perform a basic setup.

You can use `nix-build nixos/release.nix -A manual.x86_64-linux && firefox result/share/doc/nixos/index.html` to generate the doc.

flokli commented on nixos/modules/services/networking/prosody.xml:
```diff
@@ -0,0 +1,88 @@
+    <link xlink:href="https://xmpp.org/extensions/xep-0423.html">XEP-0423</link>.
+  </para>
+  <para>
+    The NixOS Prosody module will implement most of these recommandend XEPs out of
```
```suggestion
    The NixOS Prosody module will implement most of these recommended XEPs out of
```

flokli commented on nixos/modules/services/networking/prosody.xml:
```diff
@@ -0,0 +1,88 @@
+    the box. That being said, two components still require some
+    manual configuration: the
+    <link xlink:href="https://xmpp.org/extensions/xep-0045.html">Multi User Chat (MUC)</link>
+    and the <link xlink:href="https://xmpp.org/extensions/xep-0363.html">Http File Upload</link> ones.
```
```suggestion
    and the <link xlink:href="https://xmpp.org/extensions/xep-0363.html">HTTP File Upload</link> ones.
```

flokli commented on nixos/modules/services/networking/prosody.xml:
```diff
@@ -0,0 +1,88 @@
+    manual configuration: the
+    <link xlink:href="https://xmpp.org/extensions/xep-0045.html">Multi User Chat (MUC)</link>
+    and the <link xlink:href="https://xmpp.org/extensions/xep-0363.html">Http File Upload</link> ones.
+    You'll need to create a DNS subdomain for each of those. The current convention being to name your
```
```suggestion
    You'll need to create a DNS subdomain for each of those. The current convention is to name your
```

flokli commented on nixos/modules/services/networking/prosody.xml:
```diff
@@ -0,0 +1,88 @@
+    <link xlink:href="https://xmpp.org/extensions/xep-0045.html">Multi User Chat (MUC)</link>
+    and the <link xlink:href="https://xmpp.org/extensions/xep-0363.html">Http File Upload</link> ones.
+    You'll need to create a DNS subdomain for each of those. The current convention being to name your
+    MUC endpoint <literal>conference.your.domain</literal> and your HTTP upload domain <literal>upload.your.domain</literal>.
```
```suggestion
    MUC endpoint <literal>conference.example.org</literal> and your HTTP upload domain <literal>upload.example.org</literal>.
```

flokli commented on nixos/modules/services/networking/prosody.xml:
```diff
@@ -0,0 +1,88 @@
+  <para>
+    A good configuration to start with, including a
+    <link xlink:href="https://xmpp.org/extensions/xep-0045.html">Multi User Chat (MUC)</link>
+    endpoint as well as a <link xlink:href="https://xmpp.org/extensions/xep-0363.html">Http File Upload</link>
```
```suggestion
    endpoint as well as a <link xlink:href="https://xmpp.org/extensions/xep-0363.html">HTTP File Upload</link>
```

flokli commented:
Thanks a lot for all the nice work in here! :+1:
